File: python/rag/rag_stream_protocol.py
import json
from dataclasses import dataclass
from typing import Any, Dict

PROTOCOL_VERSION = 1

@dataclass
class RagStreamRequest:
    request_id: str
    query: str

@dataclass
class RagStreamEvent:
    type: str
    request_id: str
    sequence: int
    delta: str
    answer: str
    backend: str
    llm_backend: str
    error: str
    elapsed_ms: float
    finished: bool

def decode_rag_stream_request(
    message: str,
) -> RagStreamRequest:
    try:
        data: Dict[str, Any] = json.loads(message)
    except json.JSONDecodeError as exc:
        raise ValueError(
            f"invalid RAG request JSON: {exc}"
        ) from exc

    if not isinstance(data, dict):
        raise ValueError(
            "RAG stream request must be an object"
        )

    if data.get("type") != "rag_query":
        raise ValueError(
            "unsupported RAG request type"
        )

    if data.get("stream") is not True:
        raise ValueError(
            "RAG stream request must set stream=true"
        )

    if data.get("version") != PROTOCOL_VERSION:
        raise ValueError("unsupported RAG protocol version")

    request_id = data.get("request_id")
    query = data.get("query")

    if (
        not isinstance(request_id, str) or
        not request_id.strip()
    ):
        raise ValueError(
            "missing RAG request_id"
        )

    if (
        not isinstance(query, str) or
        not query.strip()
    ):
        raise ValueError(
            "missing RAG query"
        )

    return RagStreamRequest(
        request_id=request_id,
        query=query.strip(),
    )

def encode_rag_stream_event(
    event: RagStreamEvent,
) -> str:
    if not event.request_id:
        raise ValueError(
            "RAG event request_id must not be empty"
        )

    if event.sequence < 0:
        raise ValueError(
            "RAG event sequence must not be negative"
        )

    if (
        event.type == "rag_chunk" and
        (
            not event.delta or
            event.finished
        )
    ):
        raise ValueError(
            "invalid RAG chunk event"
        )

    if (
        event.type == "rag_finished" and
        (
            not event.answer or
            not event.finished
        )
    ):
        raise ValueError(
            "invalid RAG finished event"
        )

    if (
        event.type == "rag_error" and
        (
            not event.error or
            not event.finished
        )
    ):
        raise ValueError(
            "invalid RAG error event"
        )

    if event.type not in {
        "rag_chunk",
        "rag_finished",
        "rag_error",
    }:
        raise ValueError(
            f"unsupported RAG event: {event.type}"
        )

    return json.dumps(
        {
            "version": PROTOCOL_VERSION,
            "type": event.type,
            "ok": event.type != "rag_error",
            "request_id": event.request_id,
            "sequence": event.sequence,
            "delta": event.delta,
            "answer": event.answer,
            "backend": event.backend,
            "llm_backend": event.llm_backend,
            "error": event.error,
            "elapsed_ms": event.elapsed_ms,
            "finished": event.finished,
        },
        ensure_ascii=False,
    )
    

File: src/rag_stream_protocol.cpp
#include "rag_stream_protocol.h"

#include <stdexcept>
#include <string>

#include <nlohmann/json.hpp>

namespace {

constexpr const char* kQueryType = "rag_query";
constexpr const char* kChunkType = "rag_chunk";
constexpr const char* kFinishedType = "rag_finished";
constexpr const char* kErrorType = "rag_error";

using Json = nlohmann::json;

void requireObject(const Json& message) {
    if (!message.is_object()) {
        throw std::runtime_error("RAG protocol message must be a JSON object");
    }
}

void requireVersion(const Json& message) {
    if (
        !message.contains("version") ||
        !message["version"].is_number_integer() ||
        message["version"].get<int>() != RagStreamProtocol::KVersion
    ) {
        throw std::runtime_error("unsupported RAG protocol version");
    }
}

std::string requireNonEmptyString(const Json& message, const char* field) {
    if (
        !message.contains(field) ||
        !message[field].is_string()
    ) {
        throw std::runtime_error(
            std::string("missing or invalid RAG field: ") + field
        );
    }

    const std::string value = message[field].get<std::string>();

    if (value.empty()) {
        throw std::runtime_error(std::string("RAG field must not be empty: ") + field);
    }

    return value;
}

std::string optionalString(const Json& message, const char* field) {
    if (!message.contains(field)) {
        throw std::runtime_error(std::string("invalid RAG string field: ") + field);
    }

    return message[field].get<std::string>();
}

bool requireBoolean(const Json& message, const char* field) {
    if (
        !message.contains(field) ||
        !message[field].is_boolean()
    ) {
        throw std::runtime_error(std::string("missing or invalid RAG boolean field: ") + field);
    }

    return message[field].get<bool>();
}

std::size_t requireSequence(const Json& message) {
    if (
        !message.contains("sequence") ||
        !message["sequence"].is_number_unsigned()
    ) {
        throw std::runtime_error("RAG sequence must be a non-negative integer");
    }

    return message["sequence"].get<std::size_t>();
}

double requireElapsedMilliseconds(const Json& message) {
    if (
        !message.contains("elapsed_ms") ||
        !message["elapsed_ms"].is_number()
    ) {
        throw std::runtime_error("RAG elapsed_ms must be numeric");
    }

    const double elapsed = message["elapsed_ms"].get<double>();

    if (elapsed < 0.0) {
        throw std::runtime_error("RAG elapsed_ms must not be negative");
    }

    return elapsed;
}

RagStreamEventType eventTypeFromString(const std::string& type) {
    if (type == kChunkType) {
        return RagStreamEventType::Chunk;
    }

    if (type == kFinishedType) {
        return RagStreamEventType::Finished;
    }

    if (type == kErrorType) {
        return RagStreamEventType::Error;
    }

    throw std::runtime_error("unsupported RAG stream event type: " + type);
}

void validateEvent(const RagStreamEvent& event) {
    switch (event.type) {
        case RagStreamEventType::Chunk:
            if (
                !event.ok ||
                event.delta.empty() ||
                event.finished
            ) {
                throw std::runtime_error("invalid RAG chunk event");
            }
            return;

        case RagStreamEventType::Finished:
            if (
                !event.ok ||
                event.answer.empty() ||
                !event.finished
            ) {
                throw std::runtime_error("invalid RAG finished event");
            }
            return;

        case RagStreamEventType::Error:
            if (
                event.ok ||
                event.error.empty() ||
                !event.finished
            ) {
                throw std::runtime_error("invalid RAG error event");
            }
            return;
    }

    throw std::runtime_error("invalid RAG stream event");
}

}   // namespace

std::string RagStreamProtocol::encodeRequest(const RagStreamRequest& request) {
    if (request.request_id.empty()) {
        throw std::invalid_argument("RAG request_id must not be empty");
    }
    if (request.query.empty()) {
        throw std::invalid_argument("RAG query must not be empty");
    }

    const Json message = {
        {"version", KVersion},
        {"type", kQueryType},
        {"request_id", request.request_id},
        {"query", request.query},
        {"stream", true}
    };

    return message.dump();
}

RagStreamRequest RagStreamProtocol::decodeRequest(const std::string& message) {
    const Json json = Json::parse(message);

    requireObject(json);
    requireVersion(json);

    if (
        !json.contains("type") ||
        !json["type"].is_string() ||
        json["type"].get<std::string>() != kQueryType
    ) {
        throw std::runtime_error("unsupported RAG request type");
    }

    if (
        !json.contains("stream") ||
        !json["stream"].is_boolean() ||
        !json["stream"].get<bool>()
    ) {
        throw std::runtime_error("RAG stream request must set stream=true");
    }

    RagStreamRequest request;

    request.request_id = requireNonEmptyString(json, "request_id");
    request.query = requireNonEmptyString(json, "query");

    return request;
}

std::string RagStreamProtocol::encodeEvent(const RagStreamEvent& event) {
    validateEvent(event);

    if (event.request_id.empty()) {
        throw std::invalid_argument("RAG event request_id must not be empty");
    }
    if (event.elapsed_ms < 0.0) {
        throw std::invalid_argument("RAG event elapsed_ms must not be negative");
    }

    const Json message = {
        {"version", KVersion},
        {"type", eventTypeToString(event.type)},
        {"ok", event.ok},
        {"request_id", event.request_id},
        {"sequence", event.sequence},
        {"delta", event.delta},
        {"answer", event.answer},
        {"backend", event.backend},
        {"llm_backend", event.llm_backend},
        {"error", event.error},
        {"elapsed_ms", event.elapsed_ms},
        {"finished", event.finished}
    };

    return message.dump();
}

RagStreamEvent RagStreamProtocol::decodeEvent(const std::string& message) {
    const Json json = Json::parse(message);

    requireObject(json);
    requireVersion(json);

    const std::string type = requireNonEmptyString(json, "type");

    RagStreamEvent event;

    event.type = eventTypeFromString(type);
    event.ok = requireBoolean(json, "ok");

    event.request_id = requireNonEmptyString(json, "request_id");

    event.sequence = requireSequence(json);

    event.delta = optionalString(json, "delta");
    event.answer = optionalString(json, "answer");

    event.backend = optionalString(json, "backend");
    event.llm_backend = optionalString(json, "llm_backend");

    event.error = optionalString(json, "error");

    event.elapsed_ms = requireElapsedMilliseconds(json);

    event.finished = requireBoolean(json, "finished");

    validateEvent(event);

    return event;
}

std::string RagStreamProtocol::eventTypeToString(RagStreamEventType type) {
    switch (type) {
        case RagStreamEventType::Chunk:
            return kChunkType;

        case RagStreamEventType::Finished:
            return kFinishedType;

        case RagStreamEventType::Error:
            return kErrorType;
    }

    throw std::runtime_error("unsupported RAG stream event type");
}
